### PR TITLE
Party Put Post 

### PR DESCRIPTION
**Pull request by MatthewBarghout**
Motivation

  This PR implements the Create and Update endpoints for party registrations,
  allowing both students and admins to register and modify parties through the
  API. These endpoints complete the CRUD operations for the party module
  alongside the existing GET and DELETE routes.

  Changes

  Added DTOs for role-based party creation:
  - Created StudentCreatePartyDTO and AdminCreatePartyDTO to handle different
  permission levels. Students can only set themselves as contact_one, while
  admins can specify both contacts explicitly.

  Implemented POST /api/parties endpoint:
  - Added route to create new party registrations with authentication for both
  students and admins
  - Automatically creates location records in the database if the provided
  Google Maps place_id doesn't exist yet

  Implemented PUT /api/parties/{party_id} endpoint:
  - Added route to update existing party registrations
  - Maintains the same role-based permission model as the POST endpoint

  Added supporting service methods:
  - Created get_or_create_location() to handle location lookup and creation from
   Google Maps API
  - Implemented create_party_from_student_dto() and
  create_party_from_admin_dto() for role-specific party creation
  - Implemented update_party_from_student_dto() and
  update_party_from_admin_dto() for role-specific updates

  Added comprehensive test coverage:
  - Wrote tests for both POST and PUT endpoints covering student and admin
  scenarios
  - Added tests for validation errors and edge cases

Closes: #38

### CODE REVIEW

File: backend/src/modules/party/party_service.py [hunk 1]
```diff
@@ -64,6 +70,53 @@ async def _validate_student_exists(self, student_id: int) -> None:
         if result.scalar_one_or_none() is None:
             raise StudentNotFoundException(student_id)
 
+    async def _get_location_by_place_id(self, place_id: str) -> LocationEntity | None:
+        """Get a location from the database by Google Maps place_id."""
+        result = await self.session.execute(
+            select(LocationEntity).where(LocationEntity.google_place_id == place_id)
+        )
+        return result.scalar_one_or_none()
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: I'd like to stay consistent with our error handling and throw a relevant exception intead of returning None

File: backend/src/modules/party/party_router.py [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
+from src.modules.account.account_model import Account, AccountRole
+from .party_model import Party, PaginatedPartiesResponse, StudentCreatePartyDTO, AdminCreatePartyDTO
 from .party_service import PartyService
 
 party_router = APIRouter(prefix="/api/parties", tags=["parties"])
 
 
+@party_router.post("/")
+async def create_party(
+    party_data: dict = Body(...),
```
Comment (naasanov) [inline backend/src/modules/party/party_router.py]: Instead of parsing a dict you should make the type a union of the two create DTOs, and just validate that the roles line up

File: backend/src/modules/party/party_service.py [hunk 3]
```diff
@@ -142,6 +195,106 @@ async def update_party(self, party_id: int, data: PartyData) -> Party:
         await self.session.refresh(party_entity)
         return party_entity.to_model()
 
+    async def create_party_from_student_dto(
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: This is missing some validation steps. Refer to TDD 4.2.1 Registration Form Validation

**Review by naasanov — CHANGES_REQUESTED** [review]

**Review by naasanov — CHANGES_REQUESTED** [review]

File: backend/src/modules/party/party_service.py [hunk 4]
```diff
@@ -142,6 +352,128 @@ async def update_party(self, party_id: int, data: PartyData) -> Party:
...
+    ) -> Party:
+        """Create a party registration from a student. contact_one is auto-filled."""
+        # Validate student party prerequisites (date and Party Smart)
+        await self._validate_student_party_prerequisites(student_account_id, dto.party_datetime)
+
+        # Get/create location and validate no hold
+        location = await self._validate_and_get_location(dto.place_id)
+
+        # Get or create contact_two from ContactDTO
+        contact_two_id = await self._get_or_create_student_from_contact(dto.contact_two)
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: Missed this in the first review, I should have expanded more when I asked about updatind the DTOs, but as I mentioned in my post meeting summary last week, contact two is no longer stored as a student in the database. Within the OCSL system, the second contact is just an emergency backup, and does not have to complete the course or anything. This will require changes across the service, and updating the entity to match the database model in the TDD

File: backend/test/modules/party/party_router_test.py [hunk 5]
```diff
@@ -516,3 +603,503 @@ async def test_admin_authentication_required(
...
+        phone_number="5555555555",
+        account_id=3,
+        last_registered=None  # No Party Smart attendance
+    )
+    test_async_session.add_all([account, student])
+    await test_async_session.commit()
+
+    # Override authenticate_user to return the new student
+    from src.main import app
+    from src.core.authentication import authenticate_user
```
Comment (naasanov) [inline backend/test/modules/party/party_router_test.py]: Avoid in function imports

File: .claude/settings.local.json
Comment (naasanov) [inline .claude/settings.local.json]: Remove this

**Review by naasanov — CHANGES_REQUESTED** [review]

**Review by naasanov — APPROVED** [review]
